csrc/dynamic_transform.cpp

```diff
@@ -385,6 +648,219 @@ void DynamicTransformConcretizer::concretize() {
+  for (const auto& empty_tv_descr : info_->getEmptyTensors()) {
+    auto tv = info_->initialInfo()->lookUpTV(empty_tv_descr.tv_name);
+    auto rfactor = TensorDomain::noReductions(tv->getMaybeRFactorDomain());
+    std::vector<Val*> new_shape;
+    new_shape.reserve(rfactor.size());
+    for (auto id : rfactor) {
+      new_shape.push_back(id->getMaybeExpandedExtent());
+    }
+    for (auto ax : empty_tv_descr.empty_axes) {
+      // Hard-code zero extent for empty axes. This lets us detect empty input
+      // and output tensors during scheduling/execution.
+      registerConcretization(new_shape[ax], fusion->zeroVal());
+      new_shape[ax] = fusion->zeroVal();
+    }
+
+    auto hasEmptyRootReductionAxis = [&empty_tv_descr](TensorView* out_tv) {
+      return std::any_of(
+          empty_tv_descr.empty_axes.begin(),
+          empty_tv_descr.empty_axes.end(),
+          [&out_tv](size_t ax) {
+            return out_tv->getRootDomain().at(ax)->isReduction();
+          });
+    };
+
+    // Given a TensorView get a vector of its maybeRFactor maybeExpandedExtents
+    auto orig_shape = [](TensorView* out_tv) -> std::vector<Val*> {
+      const auto& rfactor =
+          TensorDomain::noReductions(out_tv->getMaybeRFactorDomain());
+      std::vector<Val*> out_shape;
+      out_shape.reserve(rfactor.size());
+      for (const auto id : rfactor) {
+        out_shape.push_back(id->getMaybeExpandedExtent());
+      }
+      return out_shape;
+    };
+
+    // Replace uses whose outputs might not be empty. Many expressions are
+    // guaranteed to have empty outputs if any of the inputs are empty; for
+    // example simple unary or binary ops. In those cases, we don't need to
+    // doctor the Fusion since they will have an empty tensor downstream which
+    // will cut off their dependence, resulting in those uses becoming dead
+    // code. For example, suppose we determined tv2 is empty, and we have the
+    // following Fusion:
+    //
+    //   auto tv4 = add(tv2, tv3);
+    //   fusion.addOutput(tv4);
+    //
+    // If we know that tv2 is empty in any dimension, then either tv3 has a
+    // matching empty dimension or it is broadcast in that dimension. Either
+    // way, the corresponding dimension in tv4 will be empty, so tv4 is an empty
+    // tensor. If we replace this expression with
+    //
+    //   auto tv4 = full(shape, zeroVal());
+    //
+    // Then the tensors tv2 and tv3 will become dead code if they have no other
+    // live uses. In this case tv4 is an output tensor, so we must keep it in
+    // the Fusion.
+    //
+    // Some special expressions can convert an empty tensor into a non-empty
+    // tensor; particularly pad, cat, and reduction ops. These ops might have
+    // non-empty outputs so in order to guarantee that all non- input or
+    // output tensors are removed, we need to replace those ops with an
+    // equivalent that does not have any empty inputs. For example
+    for (auto use : tv->uses()) {
```

Comment:
The control flow below looks a little convoluted. Given that in the future there could be other additional patterns where zero domains would need to be handled, I'd suggest to refactor below and make each part a separate function.

Reply:
I agree this is a good idea; I'm refactoring into an overloaded function to handle these ops....